src/bootstrap/flags.rs

```diff
@@ -239,25 +232,32 @@ Arguments:
             }
         });
 
-        // All subcommands can have an optional "Available paths" section
-        if matches.opt_present("verbose") {
-            let flags = Flags::parse(&["build".to_string()]);
-            let mut config = Config::parse(&flags.build, cfg_file.clone());
-            config.build = flags.build.clone();
-            let mut build = Build::new(flags, config);
-            metadata::build(&mut build);
-            let maybe_rules_help = step::build_rules(&build).get_help(subcommand);
-            if maybe_rules_help.is_some() {
-                extra_help.push_str(maybe_rules_help.unwrap().as_str());
-            }
-        } else {
-            extra_help.push_str(format!("Run `./x.py {} -h -v` to see a list of available paths.",
-                     subcommand).as_str());
-        }
+        // FIXME: The rustbuild rewrite does not make an available paths
+        // sections simple or easy to implement. I'm not sure how useful it is
+        // to have so perhaps we shouldn't attempt to reimplement it due to the
+        // difficulties imposed by such a move.
```

Comment:
This is actually a pretty nice feature to have from time to time, so I think it's worth exploring a little at least. Here's a random thought: would it be possible to have `impl Step for XXX` to return a vector of relevant paths? That way the path filtering logic of `should_run` could maybe be canonicalized somewhere?

Either that or we could have:

```rust
trait Step {
    fn should_run(...) -> ShouldRun;
}

impl ShouldRun {
    fn new() -> ShouldRun; // runs nothing
    fn path(self, &str) -> ShouldRun; // add a path this runs on
    // ... anything else necessary
}
```

basically just have a builder to house all sorts of configuration, and then we'd be able to weave all the builders together to generate these lists here.